src/elements/core/mixins/element-internals-mixin.ts

```diff
@@ -0,0 +1,91 @@
+import { isServer, type LitElement, type ReactiveController } from 'lit';
+
+import type { AbstractConstructor } from './constructor.js';
+
+// Most of our target browsers support the :state() pseudo class, but not all of them.
+// We patch the states property of the element internals to use attributes instead,
+// if :state() is not supported.
+// Note that for SSR, the states property is a simple Set, implemented in Lit SSR.
+// TODO: Remove in 2026.
+type CustomStateSetInterface = CustomStateSet;
+// eslint-disable-next-line @typescript-eslint/naming-convention
+const CustomStateSetPolyfill: (new (host: LitElement) => CustomStateSetInterface) | null =
```

Comment:
I think the polyfill could have some tests. but maybe you are planning that for the next PR.